lib/pf/Switch/Meraki/AP_http_V2.pm

```diff
@@ -0,0 +1,340 @@
+package pf::Switch::Meraki::AP_http_V2;
+
+=head1 NAME
+
+pf::Switch::Meraki::AP_http_V2
+
+=head1 SYNOPSIS
+
+The pf::Switch::Meraki::AP_http_V2 module implements an object oriented interface to
+manage the external captive portal on Meraki access points
+
+=head1 STATUS
+
+Developed and tested on a MR12 access point
+
+=head1 BUGS AND LIMITATIONS
+
+In the current BETA version, VLAN assignment is broken in Mac Authentication Bypass.
+You can work around this by using the following RADIUS filter (conf/radius_filters.conf)
+
+    [your_ssid]
+    filter = ssid
+    operator = is
+    value = Meraki-Mac-Auth-SSID
+
+    [open_ssid_meraki_hack:your_ssid]
+    scope = returnRadiusAccessAccept
+    merge_answer = no
+    answer1 = Airespace-ACL-Name => VLAN$vlan
+
+Then creating a policy named VLANXYZ where XYZ is the VLAN ID you want to assign.
+
+Using this, you will be able to configure the VLAN ids in PacketFence and simply disable the RADIUS filter when the issue is fixed on the Meraki controller. 
+
+=cut
+
+use strict;
+use warnings;
+
+use base ('pf::Switch');
+
+use Net::SNMP;
+use Try::Tiny;
+
+use pf::Switch::constants;
+use pf::constants;
+use pf::web::util;
+use pf::config;
+use pf::roles::custom;
+use pf::util;
+use pf::node;
+use pf::util::radius qw(perform_coa perform_disconnect);
+use pf::node qw(node_view);
+use pf::violation;
+use pf::locationlog;
+
+
+=head1 SUBROUTINES
+
+=cut
+
+# CAPABILITIES
+# access technology supported
+sub description { 'Meraki cloud controller v2' }
+sub supportsWirelessMacAuth { return $TRUE; }
+sub supportsWirelessDot1x { return $TRUE; }
+sub supportsExternalPortal { return $TRUE; }
+sub supportsRoleBasedEnforcement { return $TRUE; }
+sub supportsUrlBasedEnforcement { return $TRUE; }
+
+=head2 getVersion - obtain image version information from switch
+
+=cut
+
+sub getVersion {
+    my ($self) = @_;
+    my $logger = $self->logger;
+    $logger->info("we don't know how to determine the version through SNMP !");
+    return '1';
+}
+
+=item returnRoleAttribute
+
+What RADIUS Attribute (usually VSA) should the role returned into.
+
+=cut
+
+sub returnRoleAttribute {
+    return 'Airespace-ACL-Name';
+}
+
+=head2 deauthTechniques
+
+Return the reference to the deauth technique or the default deauth technique.
+
+=cut
+
+sub deauthTechniques {
+    my ($self, $method) = @_;
+    my $logger = $self->logger;
+    my $default = $SNMP::RADIUS;
+    my %tech = (
+        $SNMP::RADIUS => 'deauthenticateMacDefault',
+    );
+
+    if (!defined($method) || !defined($tech{$method})) {
+        $method = $default;
+    }
+    return $method,$tech{$method};
+}
+
+sub parseSwitchIdFromRequest {
+    my($class, $req) = @_;
+    return $$req->param('ap_mac');
+}
+
+=head2 returnRadiusAccessAccept
+
+Overloading L<pf::Switch>'s implementation to return specific attributes.
+
+=cut
+
+sub returnRadiusAccessAccept {
```

Comment:
There is an even crazier way use Role::Tiny.
http://perltricks.com/article/47/2013/11/11/Use-roles-in-object-oriented-Perl-with-Role--Tiny/
http://search.cpan.org/~haarg/Role-Tiny-2.000002/lib/Role/Tiny.pm